systems/plants/RigidBodyManipulator.cpp

```diff
@@ -1898,6 +1961,203 @@ void RigidBodyManipulator::forwarddJac(const int body_or_frame_id, const MatrixB
+{
+  if (gradient_order > 1) {
+    throw std::runtime_error("only first order gradients are available");
+  }
+
+  int nv = num_velocities;
+  int nq = num_dof;
+  GradientVar<Scalar, Eigen::Dynamic, Eigen::Dynamic> ret(nv, nv, nq, gradient_order);
+  ret.value().setZero();
+  if (gradient_order > 0)
+    ret.gradient().value().setZero();
+
+  updateCompositeRigidBodyInertias(gradient_order);
+
+  for (int i = 0; i < num_bodies; i++) {
+    RigidBody& body_i = *bodies[i];
+    if (body_i.hasParent()) {
+      int v_start_i = body_i.velocity_num_start;
+      int nv_i = body_i.getJoint().getNumVelocities();
+      GradientVar<Scalar, TWIST_SIZE, Eigen::Dynamic> F(TWIST_SIZE, nv_i, nq, gradient_order);
+      F.value().noalias() = (Ic_new[i] * body_i.J).eval();
+
+      // Hii
+      ret.value().block(v_start_i, v_start_i, nv_i, nv_i).noalias() = (body_i.J.transpose() * F.value()).eval();
+      if (gradient_order > 0) {
+        F.gradient().value() = matGradMultMat(Ic_new[i], body_i.J, dIc_new[i], body_i.dJdq);
+        auto dHii = matGradMultMat(body_i.J.transpose(), F.value(), transposeGrad(body_i.dJdq, TWIST_SIZE), F.gradient().value());
+        for (int row = 0; row < nv_i; row++) {
+          for (int col = 0; col < nv_i; col++) {
+            setSubMatrixGradient<Eigen::Dynamic>(ret.gradient().value(), getSubMatrixGradient<Eigen::Dynamic>(dHii, row, col, nv_i), v_start_i + row, v_start_i + col, nv);
+          }
+        }
+
+      }
+
+      // Hij
+      int j = body_i.parent;
+      while (bodies[j]->hasParent()) {
+        RigidBody& body_j = *bodies[j];
+        int v_start_j = body_j.velocity_num_start;
+        int nv_j = body_j.getJoint().getNumVelocities();
+        auto Hji = (body_j.J.transpose() * F.value()).eval();
+        ret.value().block(v_start_j, v_start_i, nv_j, nv_i) = Hji;
+        ret.value().block(v_start_i, v_start_j, nv_i, nv_j) = Hji.transpose();
+
+        if (gradient_order > 0) {
+          auto dHji = matGradMultMat(body_j.J.transpose(), F.value(), transposeGrad(body_j.dJdq, TWIST_SIZE), F.gradient().value());
+          for (int row = 0; row < Hji.rows(); row++) {
+            for (int col = 0; col < Hji.cols(); col++) {
+              auto dHji_element = getSubMatrixGradient<Eigen::Dynamic>(dHji, row, col, Hji.rows());
+              setSubMatrixGradient<Eigen::Dynamic>(ret.gradient().value(), dHji_element, row + v_start_j, col + v_start_i, nv);
+              setSubMatrixGradient<Eigen::Dynamic>(ret.gradient().value(), dHji_element, col + v_start_i, row + v_start_j, nv);
+            }
+          }
+        }
+        j = bodies[j]->parent;
+      }
+    }
+  }
+
+  return ret;
+}
+
+template <typename Scalar>
+GradientVar<Scalar, Eigen::Dynamic, 1> RigidBodyManipulator::inverseDynamics(
+    std::map<int, std::unique_ptr<GradientVar<Scalar, TWIST_SIZE, 1> > >& f_ext,
+    GradientVar<Scalar, Eigen::Dynamic, 1>* vd, int gradient_order)
```

Comment:
I mentioned this a while ago, but the algorithm for computing C is basically an inverse dynamics algorithm with the joint acceleration set to zero. I chose to add the option of passing in a nonzero joint acceleration vector, turning this back into an inverse dynamics algorithm that doesn't require computing the mass matrix.